src/cpu/flexcpu/flexcpu.cc

```diff
@@ -0,0 +1,1106 @@
+                                     Tick send_time,
+                                     SplitAccCtrlBlk* split)
+{
+    PacketPtr pkt;
+
+    memLatency.sample(curTick() - send_time);
+
+    if (split) {
+        if (orig_pkt == split->low) {
+            split->lowReceived = true;
+        } else if (orig_pkt == split->high) {
+            split->highReceived = true;
+        } else {
+            panic("Malformed split MemAccessReq received!");
+        }
+
+        // If the other half of the request was killed somehow, we need to
+        // clean up to avoid memory leaks. Only one of the conditions here will
+        // be true, because the other one will match orig_pkt, and be deleted
+        // via the recvTimingResp() function.
+        if (!split->low || !split->high) {
+            delete split->main;
+            delete split;
+            return;
+        }
+
+        if (!(split->lowReceived && split->highReceived))
+            return; // We need to wait for the other to complete as well
+
+        pkt = split->main;
+        assert(pkt);
+        pkt->makeResponse();
+    } else {
+        pkt = orig_pkt;
+    }
+
+    const shared_ptr<ExecContext> ctxt = context.lock();
+
+    Fault fault = ctxt && !inst->isStore() ?
+        inst->completeAcc(pkt, ctxt.get(), trace_data) : NoFault;
+
+    if (split) {
+        // The packets we actually sent to memory are deleted by recvTimingResp
+        delete split->main;
+        delete split;
+    }
+
+    callback(fault);
+}
+
+MasterPort&
+FlexCPU::getDataPort()
+{
+    return _dataPort;
+}
+
+MasterPort&
+FlexCPU::getInstPort()
+{
+    return _instPort;
+}
+
+void
+FlexCPU::init()
+{
+    BaseCPU::init();
+
+    if (!params()->switched_out &&
+        system->getMemoryMode() != Enums::timing)
+    {
+        fatal("The Dataflow CPU requires the memory system to be in "
+              "'timing' mode.\n");
+    }
+
+    for (auto& thread : threads) {
+        ThreadContext* tc = thread->getThreadContext();
+        tc->initMemProxies(tc);
+
+        // TODO if fullsystem, we need to check if we need to call initCPU on
+        // the ISA
+    }
+    // TODO add post-construction initialization code
+}
+
+void
+FlexCPU::markActiveCycle()
+{
+    if (curTick() != lastActiveTick) {
+        numCycles++;
+        lastActiveTick = curTick();
+        for (auto& thread : threads) {
+            thread->recordCycleStats();
+        }
+    }
+}
+
+void
+FlexCPU::requestDataAddrTranslation(const RequestPtr& req,
+    ThreadContext* tc, bool write, std::weak_ptr<ExecContext> context,
+    TranslationCallback callback_func)
+{
+    DPRINTF(FlexCPUCoreEvent, "requestDataAddrTranslation()\n");
+    // delay if we're out of slots for simultaneous dtb translations
+
+    Tick queue_time = curTick();
+
+    dataAddrTranslationUnit.addRequest([this, req, tc, write, context,
+                                        callback_func, queue_time]
+    {
```

Comment:
Do we need to use a lambda here? Can we make this function a member of a class?

I won't write this on every lambda, but we should think about it for every one.

Reply:
Well yes, if we want to use the addRequest interface function in the resource class, since a member function is being called (this).